### PR TITLE
SONARJAVA-2873 Symbolic execution should not make invalid assumptions on the result of "Optional.filter(...)"

### PR DESCRIPTION
**Pull request by christophe-zurn-sonarsource**

### CODE REVIEW

File: java-frontend/src/main/java/org/sonar/java/se/checks/OptionalGetBeforeIsPresentCheck.java [hunk 1]
```diff
@@ -130,22 +149,33 @@ public IsPresentSymbolicValue(SymbolicValue sv) {
...
       if (isImpossibleState(booleanConstraint, optionalConstraint)) {
         return Collections.emptyList();
       }
       if (optionalConstraint == OptionalConstraint.NOT_PRESENT || optionalConstraint == OptionalConstraint.PRESENT) {
         return Collections.singletonList(programState);
       }
       OptionalConstraint newConstraint = booleanConstraint.isTrue() ? OptionalConstraint.PRESENT : OptionalConstraint.NOT_PRESENT;
+
+      if (newConstraint == OptionalConstraint.PRESENT && optionalSV instanceof FilteredOptionalSymbolicValue) {
+        programState = ((FilteredOptionalSymbolicValue) optionalSV).propagatePresentConstraint(programState);
```
Comment (christophe-zurn-sonarsource) [inline java-frontend/src/main/java/org/sonar/java/se/checks/OptionalGetBeforeIsPresentCheck.java]: I'm not sure if there is a better way to do this, but I couldn't find any existing method allowing to do something similar to this `propagatePresentConstraint`

Comment (benzonico) [reply java-frontend/src/main/java/org/sonar/java/se/checks/OptionalGetBeforeIsPresentCheck.java]: Why not override the setConstraint method in OptionalSV for this purpose ?

File: java-frontend/src/main/java/org/sonar/java/se/checks/OptionalGetBeforeIsPresentCheck.java [hunk 2]
```diff
@@ -110,12 +110,34 @@ public ProgramState checkPostStatement(CheckerContext context, Tree syntaxNode)
...
+    private FilteredOptionalSymbolicValue(SymbolicValue wrappedValue) {
+      super(wrappedValue);
+    }
+
+    @Override
+    public List<ProgramState> setConstraint(ProgramState programState, Constraint constraint) {
+      ProgramState ps = programState;
+      if (constraint == OptionalConstraint.PRESENT) {
+        List<ProgramState> programStates = wrappedValue.setConstraint(ps, constraint);
+        Preconditions.checkState(programStates.size() == 1);
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/se/checks/OptionalGetBeforeIsPresentCheck.java]: This deserves a comment.

File: java-frontend/src/main/java/org/sonar/java/se/checks/OptionalGetBeforeIsPresentCheck.java [hunk 3]
```diff
@@ -130,22 +152,29 @@ public IsPresentSymbolicValue(SymbolicValue sv) {
      * Will be called only after calling Optional.isPresent()
      */
     @Override
-    public List<ProgramState> setConstraint(ProgramState programState, BooleanConstraint booleanConstraint) {
-      OptionalConstraint optionalConstraint =  programState.getConstraint(optionalSV, OptionalConstraint.class);
+    public List<ProgramState> setConstraint(ProgramState ps, BooleanConstraint booleanConstraint) {
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/se/checks/OptionalGetBeforeIsPresentCheck.java]: Unecessary change

**Pull request merged**